### PR TITLE
EtherRewrite and SpinlockPush

### PR DESCRIPTION
**Pull request by tbarbette**
A PR for two elements I wrote I re-used multiple times that may be of interest.
- EtherRewrite element :
Overwrite an ethernet source and destination address. This is usefull for a configuration where there is no bridge on the other end of a given output link, meaning that the source and destination address for all packets going through that link will be constant. Strip(14)->EntherEncap, or EnsureEther would rewrite the ethertype, which would need to explicitly take care of all possible ethertype. This can be replaced by two StoreEtherAddress but I think it is more "clean".

- SpinlockPush element for push paths : 
Acquire a lock, push a packet, then release the lock. Contrary to a SpinlockAcquire/Release duo, it supports dropping packets. It is practical if some elements between this one and the next PushToPull (or discarding element) are not MT-safe, as only one thread will traverse the downstream path at anyone time.
That was my first way to go for ToDpdkDevice. But it did meant to lock per paquet, which is far too expensive (ToDpdkDevice lock per burst). But I think this element may still be usefull in some cases as there is still many non-threadsafe elements in Click.

### CODE REVIEW

File: elements/ethernet/ensureether.hh [hunk 1]
```diff
@@ -24,19 +24,15 @@ specified by the arguments is prepended to the packet.
...
-
-=n
-
-For IP packets you probably want to use ARPQuerier instead.
+  EnsureEther(0x0800, 1:1:1:1:1:1, 2:2:2:2:2:2)
 
 =a
 
-EtherEncap */
+EnsureEther */
```
Comment (kohler) [inline elements/ethernet/ensureether.hh]: Hm, this doc field is usually used for “see also”, it's weird for an element to refer to itself.

File: elements/threads/spinlockpush.cc [hunk 2]
```diff
@@ -0,0 +1,41 @@
+/*
+ * spinlockpush.{cc,hh} -- element acquires spinlock
+ * Benjie Chen, Eddie Kohler
```
Comment (kohler) [inline elements/threads/spinlockpush.cc]: “Tom Barbette”

**Comment by tbarbette** [comment]
Done according to your comments ;)

**Comment by kohler** [comment]
Thanks!!!

**Pull request merged**